ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextShadowNode.java

```diff
@@ -62,96 +154,38 @@ public long measure(
-                      .setAlignment(alignment)
-                      .setLineSpacing(0.f, 1.f)
-                      .setIncludePad(mIncludeFontPadding)
-                      .setBreakStrategy(mTextBreakStrategy)
-                      .setHyphenationFrequency(mHyphenationFrequency);
-
-              if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
-                builder.setJustificationMode(mJustificationMode);
-              }
-              if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-                builder.setUseLineSpacingFromFallbacks(true);
-              }
-              layout = builder.build();
-            }
-
-          } else if (boring != null && (unconstrainedWidth || boring.width <= width)) {
-            // Is used for single-line, boring text when the width is either unknown or bigger
-            // than the width of the text.
-            layout =
-                BoringLayout.make(
-                    text,
-                    textPaint,
-                    boring.width,
-                    alignment,
-                    1.f,
-                    0.f,
-                    boring,
-                    mIncludeFontPadding);
-          } else {
-            // Is used for multiline, boring text and the width is known.
-
-            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
-              layout =
-                  new StaticLayout(
-                      text, textPaint, (int) width, alignment, 1.f, 0.f, mIncludeFontPadding);
-            } else {
-              StaticLayout.Builder builder =
-                  StaticLayout.Builder.obtain(text, 0, text.length(), textPaint, (int) width)
-                      .setAlignment(alignment)
-                      .setLineSpacing(0.f, 1.f)
-                      .setIncludePad(mIncludeFontPadding)
-                      .setBreakStrategy(mTextBreakStrategy)
-                      .setHyphenationFrequency(mHyphenationFrequency);
-
-              if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-                builder.setUseLineSpacingFromFallbacks(true);
+          Layout layout = measureSpannedText(text, width, widthMode);
+
+          if (mAdjustsFontSizeToFit) {
+            int initialFontSize = mTextAttributes.getEffectiveFontSize();
+            int currentFontSize = mTextAttributes.getEffectiveFontSize();
+            // Minimum font size is 4pts to match the iOS implementation.
+            int minimumFontSize = (int) Math.max(mMinimumFontScale * initialFontSize, PixelUtil.toPixelFromDIP(4));
+            while (
+              currentFontSize > minimumFontSize && (
+                mNumberOfLines != UNSET && layout.getLineCount() > mNumberOfLines ||
+                heightMode != YogaMeasureMode.UNDEFINED && layout.getHeight() > height)
+            ) {
+              // TODO: We could probably use a smarter algorithm here. This will require 0(n) measurements
```

Comment:
We could use some sort of binary search here like on iOS but I didn't do it for now since it adds quite a bit of complexity. In my tests it took between 5-10 iterations here.

Reply:
For now when this is merged in, we should probably add documentation indicating that this feature could be pretty slow. 

Is it possible for this `while` loop to never terminate? Should we enforce a max # of iterations so that we never get into an infinite loop?

Reply:
👍 

The loop will always terminate because of the `currentFontSize > minimumFontSize` condition. The minimum font size will never be smaller than 4.